crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1141,6 +1045,390 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
     }
 }
 
+/// Algorithm ensuring that day dividers are adjusted correctly, according to
+/// new items that have been inserted.
+pub(super) struct DayDividerAdjuster<'a, 'o> {
+    pub items: &'a mut ObservableVectorTransaction<'o, Arc<TimelineItem>>,
+    pub meta: &'a mut TimelineInnerMetadata,
+}
+
+#[derive(Debug)]
+enum DayDividerOperation {
+    Insert(usize, MilliSecondsSinceUnixEpoch),
+    Replace(usize, MilliSecondsSinceUnixEpoch),
+    Remove(usize),
+}
+
+impl<'a, 'o> DayDividerAdjuster<'a, 'o> {
+    /// Ensures that date separators are properly inserted/removed when needs
+    /// be.
+    #[instrument(skip(self))]
+    pub fn maybe_adjust_day_dividers(&mut self) {
+        // We're going to record vector operations like inserting, replacing and
+        // removing day dividers. Since we may remove or insert new items,
+        // recorded offsets will change as we're iterating over the array. The
+        // only way this is possible is because we're recording operations
+        // happening in non-decreasing order of the indices, i.e. we can't do an
+        // operation onindex I and then on any index J<I later.
+        //
+        // Note we can't just iterate in reverse order, because we may have a
+        // `Remove(i)` followed by a `Replace((i+1) -1)`, which wouldn't do what
+        // we want, if running in reverse order.
+
+        let mut ops = vec![];
+
+        let mut prev_item: Option<(&Arc<TimelineItem>, MilliSecondsSinceUnixEpoch)> = None;
+        let mut last_event_ts = None;
+
+        for (i, item) in self.items.iter().enumerate() {
```

Comment:
Too bad https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map_windows is still unstable. I suspect it could have made the code a little bit simpler.